pkg/controllers/user/rbac/legacy_grb_cleaner.go

```diff
@@ -56,8 +62,35 @@ func (p *grbCleaner) removeStuckFinalizer(obj *v3.GlobalRoleBinding) (*v3.Global
 
 	if modified {
 		md.SetFinalizers(finalizers)
-		obj, e := p.m.workload.Management.Management.GlobalRoleBindings("").Update(obj)
-		return obj, e
+	}
+	return obj, nil
+}
+
+func (p *grbCleaner) removeFinalizerFromNonExistentCluster(obj *v3.GlobalRoleBinding) (*v3.GlobalRoleBinding, error) {
+	obj = obj.DeepCopy()
+	modified := false
+
+	md, err := meta.Accessor(obj)
+	if err != nil {
+		return obj, err
+	}
+
+	finalizers := md.GetFinalizers()
+	for i := len(finalizers) - 1; i >= 0; i-- {
+		f := finalizers[i]
+		if strings.HasPrefix(f, lifecycle.ScopedFinalizerKey) {
+			s := strings.Split(f, "_")
+			// if cluster was not reported, its a deleted cluster and will cause the finalizer to hang in the future
+			if _, err = p.m.clusterLister.Get("", s[1]); errors.IsNotFound(err) {
+				finalizers = append(finalizers[:i], finalizers[i+1:]...)
+				modified = true
+			}
+		}
+	}
+
+	if modified {
+		md.SetFinalizers(finalizers)
+		return p.m.workload.Management.Management.GlobalRoleBindings("").Update(obj)
```

Comment:
This isn't needed, call update after you add the annotation or update is going to get called twice in a row which is a good way to get conflict errors.